File: csharp/UrakawaToolkit/media/data/utillities/SequenceStream.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;

namespace urakawa.media.data.utillities
{
	/// <summary>
	/// A <see cref="Stream"/> that supports reading from a sequence of source <see cref="Stream"/>s
	/// as if they were one.
	/// </summary>
	public class SequenceStream : Stream
	{
		private List<Stream> mSources;
		private int mCurrentIndex;

		/// <summary>
		/// Constructor supplying the sequence of source <see cref="Stream"/>s
		/// </summary>
		/// <param name="ss">
		/// The sequence of source <see cref="Stream"/>s. 
		/// Must contain at least one source <see cref="Stream"/>
		/// </param>
		public SequenceStream(IEnumerable<Stream> ss)
		{
			mSources = new List<Stream>(ss);
			if (mSources.Count == 0)
			{
				throw new exception.MethodParameterHasNoItemsException(
					"A SequenceStream must have at least one source Stream in it's sequence");
			}
			mCurrentIndex = 0;
			mSources[0].Seek(0, SeekOrigin.Begin);
		}

		/// <summary>
		/// Gets a value indicating whether the current stream supports reading. 
		/// </summary>
		public override bool CanRead
		{
			get
			{
				foreach (Stream subS in mSources)
				{
					if (!subS.CanRead)
					{
						return false;
					}
				}
				return true;
			}
		}

		/// <summary>
		/// Gets a value indicating whether the current stream supports seeking. 
		/// </summary>
		public override bool CanSeek
		{
			get
			{
				foreach (Stream subS in mSources)
				{
					if (!subS.CanSeek)
					{
						return false;
					}
				}
				return true;
			}
		}

		/// <summary>
		/// Gets a value indicating whether the current stream supports writing - always returns <c>false</c>
		/// </summary>
		public override bool CanWrite
		{
			get { return false; }
		}

		/// <summary>
		/// Does nothing since a <see cref="SubStream"/> is read-only
		/// </summary>
		public override void Flush()
		{
			//Does nothing since there is no writing
		}

		/// <summary>
		/// Gets the length in <see cref="byte"/>s of the current <see cref="SequenceStream"/>
		/// </summary>
		public override long Length
		{
			get
			{
				long len = 0;
				foreach (Stream subS in mSources)
				{
					len += subS.Length;
				}
				return len;
			}
		}

		/// <summary>
		/// Gets or sets the position within the current <see cref="SequenceStream"/>. 
		/// </summary>
		public override long Position
		{
			get
			{
				return getBytesBeforeIndex(mCurrentIndex) + mSources[mCurrentIndex].Position;
			}
			set
			{
				mCurrentIndex = 0;
				long bytesBefore = 0;
				while (mCurrentIndex < mSources.Count)
				{
					if (value<bytesBefore+mSources[mCurrentIndex].Length)
					{
						mSources[mCurrentIndex].Position = value-bytesBefore;
						return;
					}
					bytesBefore += mSources[mCurrentIndex].Length;
					mCurrentIndex++;
				}
				mCurrentIndex = mSources.Count-1;
				mSources[mCurrentIndex].Position = mSources[mCurrentIndex].Length;
			}
		}

		private long getBytesBeforeIndex(int index)
		{
			int i = 0;
			if (index >= mSources.Count) index = mSources.Count - 1;
			long bytesBefore = 0;
			while (i < index)
			{
				bytesBefore += mSources[i].Length;
				i++;
			}
			return bytesBefore;
		}


		/// <summary>
		/// Reads a sequence of bytes from the current <see cref="SequenceStream"/> and 
		/// advances the position within the stream by the number of bytes read
		/// </summary>
		/// <param name="buffer">
		/// An array of bytes. When this method returns, 
		/// the buffer contains the specified byte array with the values between 
		/// <c><paramref name="offset"/></c> and <c>(<paramref name="offset"/> + <paramref name="count"/> - 1)</c> 
		/// replaced by the bytes read from the current <see cref="SequenceStream"/>.
		/// </param>
		/// <param name="offset">
		/// The zero-based byte offset in <paramref name="buffer"/> 
		/// at which to begin storing the data read from the current <see cref="SequenceStream"/>.
		/// </param>
		/// <param name="count">The maximum number of bytes to be read from the current <see cref="SequenceStream"/>.</param>
		/// <returns>The number of <see cref="byte"/>s read</returns>
		public override int Read(byte[] buffer, int offset, int count)
		{
			if (count == 0) return 0;
			int totalBytesRead = 0;
			int bytesRead = 0;
			while (true)
			{
				if (mSources[mCurrentIndex].Position < mSources[mCurrentIndex].Length)
				{
					bytesRead = mSources[mCurrentIndex].Read(buffer, offset, count);
				}
				else
				{
					bytesRead = 0;
				}
				totalBytesRead += bytesRead;
				count -= bytesRead;
				offset += bytesRead;
				if (count == 0) break;
				if (mCurrentIndex+1 < mSources.Count)
				{
					mCurrentIndex++;
					mSources[mCurrentIndex].Position = 0;
				}
				else
				{
					break;
				}
			}
			return totalBytesRead;
		}


		/// <summary>
		/// Sets the <see cref="Position"/> within the current stream.
		/// </summary>
		/// <param name="offset">
		/// A byte <paramref name="offset"/> relative to the origin parameter.
		/// </param>
		/// <param name="origin">
		/// A value of type <see cref="SeekOrigin"/> indicating the reference point used to obtain the new position.
		/// </param>
		/// <returns>The new <see cref="Position"/> within the current stream.</returns>
		public override long Seek(long offset, SeekOrigin origin)
		{
			long newPos = Position;
			switch (origin)
			{
				case SeekOrigin.Current:
					newPos += offset;
					break;
				case SeekOrigin.Begin:
					newPos = offset;
					break;
				case SeekOrigin.End:
					newPos = Length + offset;
					break;
			}
			Position = newPos;
			return Position;
		}

		/// <summary>
		/// Sets the <see cref="Length"/> of the <see cref="SequenceStream"/>.
		/// Since a <see cref="SequenceStream"/> is read-only, 
		/// calling this method will thorw an <see cref="NotSupportedException"/>
		/// </summary>
		/// <param name="value">The new <see cref="Length"/> </param>
		public override void SetLength(long value)
		{
			throw new NotSupportedException("A SubStream is read-only so it's length can not be set");
		}

		/// <summary>
		/// 
		/// Since a <see cref="SequenceStream"/> is read only, 
		/// calling this method will thorw an <see cref="NotSupportedException"/>
		/// </summary>
		/// <param name="buffer">
		/// An array of <see cref="byte"/>s. 
		/// This method copies <paramref name="count"/> bytes from <paramref name="buffer"/> to the current stream.
		/// </param>
		/// <param name="offset">
		/// The zero-based byte offset in <paramref name="buffer"/> at which to begin copying <see cref="byte"/>s 
		/// to the current stream.
		/// </param>
		/// <param name="count">The number of <see cref="byte"/>s to be written to the current stream.</param>
		public override void Write(byte[] buffer, int offset, int count)
		{
			throw new NotSupportedException("A SubStream is read-only so it can not be written to");
		}

		/// <summary>
		/// Closes the <see cref="SequenceStream"/> including all it's source <see cref="Stream"/>s
		/// </summary>
		public override void Close()
		{
			base.Close();
			foreach (Stream subS in mSources)
			{
				subS.Close();
			}
		}
	}
}

File: java/core/src/org/daisy/urakawa/nativeapi/SequenceStream.java
package org.daisy.urakawa.nativeapi;

import java.io.IOException;
import java.util.LinkedList;
import java.util.List;

import org.daisy.urakawa.exception.MethodParameterIsNullException;

/**
 * Represents an ordered list of Streams. This class should be replaced by an
 * equivalent IStream API in the implementing language. The methods exposed here
 * mimic the System.IO.Stream C# API.
 * 
 * @stereotype Language-Dependent
 */
public class SequenceStream implements IStream
{
    private List<IStream> mSources;
    private int mCurrentIndex;

    /**
     * @param ss
     * @throws MethodParameterIsNullException
     */
    public SequenceStream(List<IStream> ss)
            throws MethodParameterIsNullException
    {
        if (ss == null)
        {
            throw new MethodParameterIsNullException();
        }
        mSources = new LinkedList<IStream>(ss);
        if (mSources.size() == 0)
        {
            throw new MethodParameterIsNullException();
        }
        mCurrentIndex = 0;
        // mSources.get(0).Seek(0, SeekOrigin.Begin);
    }

    /**
     * @return int
     */
    public int getLength()
    {
        int len = 0;
        for (IStream subS : mSources)
        {
            len += subS.getLength();
        }
        return len;
    }

    public int getPosition()
    {
        return getBytesBeforeIndex(mCurrentIndex)
                + mSources.get(mCurrentIndex).getPosition();
    }

    public void setPosition(int pos)
    {
        mCurrentIndex = 0;
        int bytesBefore = 0;
        while (mCurrentIndex < mSources.size())
        {
            if (pos < bytesBefore + mSources.get(mCurrentIndex).getLength())
            {
                mSources.get(mCurrentIndex).setPosition(pos - bytesBefore);
                return;
            }
            bytesBefore += mSources.get(mCurrentIndex).getLength();
            mCurrentIndex++;
        }
        mCurrentIndex = mSources.size() - 1;
        mSources.get(mCurrentIndex).setPosition(
                mSources.get(mCurrentIndex).getLength());
    }

    private int getBytesBeforeIndex(int index)
    {
        int i = 0, index_ = index;
        if (index_ >= mSources.size())
            index_ = mSources.size() - 1;
        int bytesBefore = 0;
        while (i < index_)
        {
            bytesBefore += mSources.get(i).getLength();
            i++;
        }
        return bytesBefore;
    }

    public int read(byte[] buffer, int offset, int count) throws IOException
    {
        int offset_ = offset;
        int count_ = count;
        if (count_ == 0)
            return 0;
        int totalBytesRead = 0;
        int bytesRead = 0;
        while (true)
        {
            if (mSources.get(mCurrentIndex).getPosition() < mSources.get(
                    mCurrentIndex).getLength())
            {
                bytesRead = mSources.get(mCurrentIndex).read(buffer, offset_,
                        count_);
            }
            else
            {
                bytesRead = 0;
            }
            totalBytesRead += bytesRead;
            count_ -= bytesRead;
            offset_ += bytesRead;
            if (count_ == 0)
                break;
            if (mCurrentIndex + 1 < mSources.size())
            {
                mCurrentIndex++;
                mSources.get(mCurrentIndex).setPosition(0);
            }
            else
            {
                break;
            }
        }
        return totalBytesRead;
    }

    public void close() throws IOException
    {
        for (IStream subS : mSources)
        {
            subS.close();
        }
    }

    public void seek(int n)
    {
        /**
         * To implement.
         */
    }

    public void write(byte[] buffer, int offset, int count) throws IOException
    {
        /**
         * To implement.
         */
    }

    public byte readByte()
    {
        return 0;
    }

    public byte[] readBytes(int length)
    {
        return null;
    }
}
